flink-batch-connectors/flink-jdbc/src/main/java/org/apache/flink/api/java/io/jdbc/JDBCInputFormat.java

```diff
@@ -157,14 +216,25 @@ public boolean reachedEnd() throws IOException {
 	 * @throws java.io.IOException
 	 */
 	@Override
-	public OUT nextRecord(OUT tuple) throws IOException {
+	public Row nextRecord(Row row) throws IOException {
 		try {
-			resultSet.next();
-			if (columnTypes == null) {
-				extractTypes(tuple);
+			hasNext = resultSet.next();
+			if (!hasNext) {
+				return null;
+			}
+			try {
+				//This throws a NPE when the TypeInfo is not passed to the InputFormat,
+				//i.e. KryoSerializer used to generate the passed row
+				row.productArity();
```

Comment:
OK, I see. How about we extend the InputFormat to implement the `ResultTypeQueryable` interface and let users either specify all field types or at least the number of result attributes via the `JDBCInputFormatBuilder`.
Then we do not have to fall back to the KryoSerializer that does creates corrupt `Row` objects and users do not have to specify types in the `env.createInput()` method.